api/src/main/java/org/pmiops/workbench/api/UserMetricsController.java

```diff
@@ -176,22 +180,26 @@ public void setDistinctWorkspaceLimit(int limit) {
     // RW-1298
     // This needs to be refactored to only use namespace and FC ID
     // The purpose of this Map, is to check what is actually still present in FC
-    Map<Long, WorkspaceResponse> workspaceAccessMap =
-        workspaceIdList.stream()
-            .collect(
-                Collectors.toMap(
-                    id -> id,
-                    id -> {
-                      DbWorkspace workspace = workspaceService.findByWorkspaceId(id);
-                      WorkspaceResponse workspaceResponse =
-                          fireCloudService.getWorkspace(
-                              workspace.getWorkspaceNamespace(), workspace.getFirecloudName());
-                      return workspaceResponse;
-                    }));
+    ImmutableMap.Builder<Long, WorkspaceResponse> liveWorkspacesByIdBuilder = new Builder<>();
+    for (long workspaceId : workspaceIdList) {
+      final Optional<DbWorkspace> workspaceMaybe = workspaceService.getWorkspaceMaybe(workspaceId);
+      if (!workspaceMaybe.isPresent()) {
+        logger.log(
+            Level.INFO,
+            String.format("Workspace ID %d still in recent list but not found", workspaceId));
+        continue;
+      }
+      final DbWorkspace dbWorkspace = workspaceMaybe.get();
+      final WorkspaceResponse workspaceResponse = fireCloudService.getWorkspace(dbWorkspace);
+
+      Optional.ofNullable(workspaceResponse)
+          .ifPresent(r -> liveWorkspacesByIdBuilder.put(workspaceId, r));
```

Comment:
Can we keep the functional style by using a filter? I believe Optionals supports this easily.

Reply:
Yes, we definitely can, and i just happened to do it before reading this. I originally didn't do it all in a huge stream because there were exceptions in play. It is a bit to chew on if you're not familiar with the constructs.